bigtable/client/row_set.cc

```diff
@@ -0,0 +1,37 @@
+// Copyright 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "bigtable/client/row_set.h"
+
+namespace bigtable {
+inline namespace BIGTABLE_CLIENT_NS {
+namespace btproto = ::google::bigtable::v2;
+
+void RowSet::Intersect(bigtable::RowRange const& range) {
+  btproto::RowSet result;
+  for (auto const& key : row_set_.row_keys()) {
+    if (range.Contains(key)) {
+      *result.add_row_keys() = key;
+    }
+  }
+  for (auto const& r : row_set_.row_ranges()) {
+    auto i = range.Intersect(RowRange(r));
+    if (std::get<0>(i)) {
+      *result.add_row_ranges() = std::get<1>(i).as_proto_move();
```

Comment:
I guess the way `RowRange::Intersect` is implemented forces this function to copy everything even when the intersection is a no-op. Maybe `Rowset::Intersect` would be better off staying const and returning the new RowSet?

Reply:
Done.
We can always change this if we find that it is too slow (I doubt it, restarting a `ReadRows()` request should be rare).

Reply:
Thanks! I also agree it's not worth optimizing, but since this is in the public headers, having the two classes behave similarly makes them nicer.